src/app/modules/spcp/sp.oidc.service.ts

```diff
@@ -0,0 +1,310 @@
+    const jwtName = JwtName.SP
+    const cookie = cookies[jwtName]
+    return cookie ? ok(cookie) : err(new MissingJwtError())
+  }
+
+  /**
+   * Verifies a Singpass JWT and extracts its payload.
+   * @param jwt The contents of the JWT cookie
+   * @returns ok(Singpass JWT payload)
+   * @returns err(VerifyJwtError) if JWT verification failed
+   * @returns err(InvalidJwtError) if JWT has invalid shape
+   */
+  extractSingpassJwtPayload(
+    jwt: string,
+  ): ResultAsync<
+    SingpassJwtPayloadFromCookie,
+    VerifyJwtError | InvalidJwtError
+  > {
+    const logMeta = {
+      action: 'extractSingpassJwtPayload',
+    }
+
+    const result = ResultAsync.fromPromise(
+      this.#spOidcClient.verifyJwt(jwt),
+      (error) => {
+        logger.error({
+          message: 'Failed to verify JWT with auth client',
+          meta: logMeta,
+          error,
+        })
+        return new VerifyJwtError()
+      },
+    ).andThen((payload) => {
+      if (isSingpassJwtPayload(payload)) {
+        return okAsync(payload)
+      }
+      const payloadIsDefined = !!payload
+      const payloadKeys =
+        typeof payload === 'object' && !!payload && Object.keys(payload)
+      logger.error({
+        message: 'JWT has incorrect shape',
+        meta: {
+          ...logMeta,
+          payloadIsDefined,
+          payloadKeys,
+        },
+      })
+      return errAsync(new InvalidJwtError())
+    })
+
+    return result
+  }
+
+  /**
+   * Gets the sp session info from the cookies
+   * @param cookies The sp cookies set by the redirect
+   * @return ok(jwtPayload) if successful
+   * @return err(MissingJwtError) if the specified cookie does not exist
+   * @return err(VerifyJwtError) if the jwt exists but could not be authenticated
+   * @return err(InvalidJwtError) if the jwt exists but the payload is invalid
+   */
+  extractJwtPayloadFromRequest(
+    cookies: SpcpCookies,
+  ): ResultAsync<
+    JwtPayloadFromCookie,
+    VerifyJwtError | InvalidJwtError | MissingJwtError
+  > {
+    return this.extractJwt(cookies).asyncAndThen((jwtResult) => {
+      return this.extractSingpassJwtPayload(jwtResult)
+    })
+  }
+
+  /**
+   * Method to exchange auth code for decrypted and verified idToken and then extract NRIC
+   * @param code authorisation code
+   * @returns okAsync(nric)
+   * @returns errAsync(InvalidIdTokenError) if failed to retrieve NRIC
+   */
+  exchangeAuthCodeAndRetrieveNric(
+    code: string,
+  ): ResultAsync<string, InvalidIdTokenError> {
+    const logMeta = {
+      action: 'exchangeAuthCodeAndRetrieveNric',
+    }
+
+    return ResultAsync.fromPromise(
+      this.#spOidcClient
+        .exchangeAuthCodeAndDecodeVerifyToken(code)
+        .then((decodedVerifiedToken) => {
+          return this.#spOidcClient.extractNricFromIdToken(decodedVerifiedToken)
+        })
+        .then((result) => {
+          if (result instanceof Error) {
+            return Promise.reject(result)
+          }
+          return result
```

Comment:
comment: should we just use neverthrow in `extractNricFromIdToken` directly instead of this?

Reply:
thanks! as above, would not use neverthrow, in order to keep the client standalone